free_fleet/src/manager/RobotInfo.cpp

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <free_fleet/manager/RobotInfo.hpp>
+#include <free_fleet/messages/NavigationRequest.hpp>
+#include <free_fleet/messages/RelocalizationRequest.hpp>
+
+#include "internal_RobotInfo.hpp"
+#include "../manager/utilities/utilities.hpp"
+#include "../manager/requests/SimpleRequestInfo.hpp"
+
+namespace free_fleet {
+namespace manager {
+
+//==============================================================================
+void RobotInfo::Implementation::allocate_task(
+  const std::shared_ptr<RequestInfo>& new_request_info)
+{
+  bool request_registered = allocated_requests.insert({
+        new_request_info->id(),
+        new_request_info}).second;
+  if (!request_registered)
+  {
+    ffwarn << "Attempted to allocate new task with existing task ID ["
+           << new_request_info->id() << "], ignoring.\n";
+  }
+}
+
+//==============================================================================
+void RobotInfo::Implementation::update_state(
+  RobotInfo& robot_info,
+  const messages::RobotState& new_state,
+  rmf_traffic::Time time_now)
+{
+  if (robot_info.name() != new_state.name())
+    return;
+
+  const std::optional<uint32_t> task_id = new_state.task_id();
+  auto it = robot_info._pimpl->allocated_requests.end();
+  std::pair<TrackingState, std::size_t> new_tracking_estimate;
+
+  // Robot is not doing any task at the moment.
+  if (!task_id.has_value())
+  {
+    new_tracking_estimate =
+      robot_info._pimpl->track_through_graph(new_state);
+  }
+  // Use the implemented track_robot function of each request type to get new
+  // tracking estimates.
+  else if ((it = robot_info._pimpl->allocated_requests.find(task_id.value()))
```

Comment:
Minor nitpick: I'm a tiny bit uncomfortable with how much is being loaded up in this if-condition.

What if you restructure it like this:

```
if (!task_id.has_value())
{
  ...
}
else
{
  const auto it = robot_info._pimpl->allocated_requests.find(task_id.value());
  if (it != robot_info._pimpl->allocated_requests.end())
  {
    ...
  }
  else
  {
    ...
  }
}
```

Then you no longer need to forward-declare `auto it`.